File: src/kalman.py
"""
Kalman filter logic for weight processing.
"""

from datetime import datetime
from typing import Dict, Tuple, Optional, Any
import numpy as np
from pykalman import KalmanFilter

try:
    from .constants import KALMAN_DEFAULTS
except ImportError:
    from constants import KALMAN_DEFAULTS


class KalmanFilterManager:
    """Manages Kalman filter operations for weight processing."""

    @staticmethod
    def initialize_immediate(
        weight: float,
        timestamp: datetime,
        kalman_config: dict,
        observation_covariance: Optional[float] = None
    ) -> Dict[str, Any]:
        """Initialize Kalman filter immediately with first measurement."""
        initial_variance = kalman_config.get("initial_variance", KALMAN_DEFAULTS['initial_variance'])

        # Use passed observation_covariance if provided, otherwise use config value
        obs_cov = observation_covariance if observation_covariance is not None else kalman_config.get("observation_covariance", KALMAN_DEFAULTS['observation_covariance'])

        kalman_params = {
            'initial_state_mean': [weight, 0],
            'initial_state_covariance': [[initial_variance, 0], [0, 0.001]],
            'transition_covariance': [
                [kalman_config.get("transition_covariance_weight", KALMAN_DEFAULTS['transition_covariance_weight']), 0],
                [0, kalman_config.get("transition_covariance_trend", KALMAN_DEFAULTS['transition_covariance_trend'])]
            ],
            'observation_covariance': [[obs_cov]],
        }

        return {
            'kalman_params': kalman_params,
            'last_state': np.array([[weight, 0]]),
            'last_covariance': np.array([[[initial_variance, 0], [0, 0.001]]]),
            'last_timestamp': timestamp,
            'last_raw_weight': weight,
        }

    @staticmethod
    def update_state(
        state: Dict[str, Any],
        weight: float,
        timestamp: datetime,
        source: str,
        processing_config: dict,
        observation_covariance: Optional[float] = None
    ) -> Dict[str, Any]:
        """Update Kalman filter state with new measurement."""
        time_delta_days = 1.0
        if state.get('last_timestamp'):
            last_timestamp = state['last_timestamp']
            if isinstance(last_timestamp, str):
                last_timestamp = datetime.fromisoformat(last_timestamp)
            delta = (timestamp - last_timestamp).total_seconds() / 86400.0
            time_delta_days = max(0.1, min(30.0, delta))

        kalman_params = state['kalman_params']

        # Use passed observation_covariance if provided, otherwise use stored value
        obs_cov = observation_covariance if observation_covariance is not None else kalman_params['observation_covariance'][0][0]

        kalman = KalmanFilter(
            transition_matrices=np.array([[1, time_delta_days], [0, 1]]),
            observation_matrices=np.array([[1, 0]]),
            initial_state_mean=np.array(kalman_params['initial_state_mean']),
            initial_state_covariance=np.array(kalman_params['initial_state_covariance']),
            transition_covariance=np.array(kalman_params['transition_covariance']),
            observation_covariance=np.array([[obs_cov]]),
        )

        observation = np.array([[weight]])

        if state.get('last_state') is None:
            filtered_state_means, filtered_state_covariances = kalman.filter(observation)
            new_last_state = filtered_state_means
            new_last_covariance = filtered_state_covariances
        else:
            last_state = state['last_state']
            last_covariance = state['last_covariance']

            if len(last_state.shape) > 1:
                current_state = last_state[-1]
                current_covariance = last_covariance[-1]
            else:
                current_state = last_state
                current_covariance = last_covariance[0] if len(last_covariance.shape) > 2 else last_covariance

            filtered_state_mean, filtered_state_covariance = kalman.filter_update(
                current_state,
                current_covariance,
                observation=observation[0],
            )

            if len(last_state.shape) == 1:
                new_last_state = np.array([last_state, filtered_state_mean])
                new_last_covariance = np.array([last_covariance, filtered_state_covariance])
            else:
                new_last_state = np.array([last_state[-1], filtered_state_mean])
                new_last_covariance = np.array([last_covariance[-1], filtered_state_covariance])

        state['last_state'] = new_last_state
        state['last_covariance'] = new_last_covariance
        state['last_timestamp'] = timestamp
        state['last_raw_weight'] = weight

        return state

    @staticmethod
    def calculate_confidence(normalized_innovation: float) -> float:
        """Calculate confidence using smooth exponential decay."""
        alpha = 0.5
        return np.exp(-alpha * normalized_innovation ** 2)

    @staticmethod
    def create_result(
        state: Dict[str, Any],
        weight: float,
        timestamp: datetime,
        source: str,
        accepted: bool,
        observation_covariance: Optional[float] = None
    ) -> Optional[Dict[str, Any]]:
        """Create result dictionary from current state."""
        if state.get('last_state') is None:
            return None

        last_state = state['last_state']
        if len(last_state.shape) > 1:
            current_state = last_state[-1]
        else:
            current_state = last_state

        filtered_weight = current_state[0]
        trend = current_state[1]

        innovation = weight - filtered_weight

        last_covariance = state.get('last_covariance')
        current_covariance = None
        if last_covariance is not None:
            if len(last_covariance.shape) > 2:
                current_covariance = last_covariance[-1]
            else:
                current_covariance = last_covariance

            # Use passed observation_covariance if provided, otherwise use stored value
            if observation_covariance is not None:
                obs_covariance = observation_covariance
            else:
                obs_covariance = state['kalman_params']['observation_covariance'][0][0]
            innovation_variance = current_covariance[0, 0] + obs_covariance
            normalized_innovation = (
                abs(innovation) / np.sqrt(innovation_variance)
                if innovation_variance > 0
                else 0
            )
        else:
            normalized_innovation = 0

        confidence = KalmanFilterManager.calculate_confidence(normalized_innovation)

        # Calculate confidence intervals (±2σ)
        if current_covariance is not None:
            confidence_interval = 2.0 * np.sqrt(current_covariance[0, 0])
            kalman_upper = filtered_weight + confidence_interval
            kalman_lower = filtered_weight - confidence_interval
            kalman_variance = float(current_covariance[0, 0])
        else:
            kalman_upper = filtered_weight
            kalman_lower = filtered_weight
            kalman_variance = None

        # Calculate prediction error
        prediction_error = innovation if accepted else None

        return {
            "timestamp": timestamp,
            "raw_weight": weight,
            "filtered_weight": float(filtered_weight),
            "trend": float(trend),
            "trend_weekly": float(trend * 7),
            "accepted": accepted,
            "confidence": float(confidence),
            "innovation": float(innovation),
            "normalized_innovation": float(normalized_innovation),
            "source": source,
            "kalman_confidence_upper": float(kalman_upper),
            "kalman_confidence_lower": float(kalman_lower),
            "kalman_variance": kalman_variance,
            "prediction_error": float(prediction_error) if prediction_error is not None else None,
        }

    @staticmethod
    def get_current_state_values(state: Dict[str, Any]) -> Tuple[Optional[float], Optional[float]]:
        """Extract current weight and trend from state."""
        if state.get('last_state') is None:
            return None, None

        last_state = state['last_state']
        if len(last_state.shape) > 1:
            current_state = last_state[-1]
        else:
            current_state = last_state

        return float(current_state[0]), float(current_state[1])

    @staticmethod
    def calculate_time_delta_days(
        current_timestamp: datetime,
        last_timestamp: Optional[datetime]
    ) -> float:
        """Calculate days between measurements."""
        if last_timestamp is None:
            return 1.0

        if isinstance(last_timestamp, str):
            last_timestamp = datetime.fromisoformat(last_timestamp)

        delta = (current_timestamp - last_timestamp).total_seconds() / 86400.0
        return max(0.1, delta)
    
    @staticmethod
    def estimate_trend_from_buffer(measurements: list, pre_gap_trend: Optional[float] = None) -> float:
        """Estimate trend from buffered measurements."""
        if len(measurements) < 2:
            return pre_gap_trend if pre_gap_trend is not None else 0.0
        
        weights = [m['weight'] for m in measurements]
        timestamps = [datetime.fromisoformat(m['timestamp']) for m in measurements]
        
        if len(weights) == 2:
            time_diff_days = (timestamps[1] - timestamps[0]).total_seconds() / 86400.0
            if time_diff_days > 0:
                trend = (weights[1] - weights[0]) / time_diff_days
            else:
                trend = 0.0
        else:
            time_diffs = []
            weight_diffs = []
            for i in range(1, len(weights)):
                time_diff = (timestamps[i] - timestamps[i-1]).total_seconds() / 86400.0
                if time_diff > 0:
                    time_diffs.append(time_diff)
                    weight_diffs.append(weights[i] - weights[i-1])
            
            if time_diffs:
                trends = [wd/td for wd, td in zip(weight_diffs, time_diffs)]
                trend = np.median(trends)
            else:
                trend = 0.0
        
        trend = max(-0.5, min(0.5, trend))
        
        if pre_gap_trend is not None:
            trend = 0.7 * trend + 0.3 * pre_gap_trend
        
        return trend
    
    @staticmethod
    def calculate_adaptive_parameters(gap_days: float, config: Dict[str, Any]) -> Dict[str, Any]:
        """Calculate adaptive Kalman parameters based on gap duration."""
        gap_factor = min(gap_days / 30, 3.0)
        
        gap_config = config.get('kalman', {}).get('gap_handling', {})
        gap_variance_multiplier = gap_config.get('gap_variance_multiplier', 2.0)
        trend_variance_multiplier = gap_config.get('trend_variance_multiplier', 20.0)
        
        base_weight_cov = config.get('kalman', {}).get('transition_covariance_weight', KALMAN_DEFAULTS['transition_covariance_weight'])
        base_trend_cov = config.get('kalman', {}).get('transition_covariance_trend', KALMAN_DEFAULTS['transition_covariance_trend'])
        
        adaptive_weight_cov = base_weight_cov * (1 + gap_factor * gap_variance_multiplier)
        adaptive_trend_cov = base_trend_cov * (1 + gap_factor * trend_variance_multiplier)
        
        return {
            'gap_factor': gap_factor,
            'transition_covariance': [
                [adaptive_weight_cov, 0],
                [0, adaptive_trend_cov]
            ]
        }
    
    @staticmethod
    def initialize_from_buffer(buffer_data: Dict[str, Any], config: Dict[str, Any]) -> Dict[str, Any]:
        """Initialize Kalman filter from buffered measurements."""
        measurements = buffer_data['measurements']
        gap_days = buffer_data['gap_days']
        
        weights = [m['weight'] for m in measurements]
        initial_weight = np.median(weights)
        
        pre_gap_trend = buffer_data.get('pre_gap_trend')
        initial_trend = KalmanFilterManager.estimate_trend_from_buffer(measurements, pre_gap_trend)
        
        adaptive_params = KalmanFilterManager.calculate_adaptive_parameters(gap_days, config)
        gap_factor = adaptive_params['gap_factor']
        
        kalman_config = config.get('kalman', {})
        initial_variance = kalman_config.get('initial_variance', KALMAN_DEFAULTS['initial_variance'])
        obs_cov = kalman_config.get('observation_covariance', KALMAN_DEFAULTS['observation_covariance'])
        
        adaptive_initial_variance = initial_variance * (1 + gap_factor * 2)
        adaptive_trend_variance = 0.001 * (1 + gap_factor * 10)
        
        kalman_params = {
            'initial_state_mean': [initial_weight, initial_trend],
            'initial_state_covariance': [
                [adaptive_initial_variance, 0],
                [0, adaptive_trend_variance]
            ],
            'transition_covariance': adaptive_params['transition_covariance'],
            'observation_covariance': [[obs_cov]],
        }
        
        state = {
            'kalman_params': kalman_params,
            'last_state': np.array([[initial_weight, initial_trend]]),
            'last_covariance': np.array([[
                [adaptive_initial_variance, 0],
                [0, adaptive_trend_variance]
            ]]),
            'gap_adaptation': {
                'active': True,
                'gap_factor': gap_factor,
                'measurements_since_gap': len(measurements),
                'initial_trend': initial_trend,
                'decay_rate': config.get('kalman', {}).get('gap_handling', {}).get('adaptation_decay_rate', 5.0)
            }
        }
        
        return state
    
    @staticmethod
    def apply_adaptation_decay(state: Dict[str, Any], config: Dict[str, Any]) -> Dict[str, Any]:
        """Apply exponential decay to adaptive parameters."""
        if not state.get('gap_adaptation', {}).get('active'):
            return state
        
        adaptation = state['gap_adaptation']
        measurements_since = adaptation['measurements_since_gap']
        
        decay_factor = np.exp(-measurements_since / adaptation['decay_rate'])
        
        if decay_factor > 0.01:
            gap_factor = adaptation['gap_factor']
            
            kalman_config = config.get('kalman', {})
            base_weight_cov = kalman_config.get('transition_covariance_weight', KALMAN_DEFAULTS['transition_covariance_weight'])
            base_trend_cov = kalman_config.get('transition_covariance_trend', KALMAN_DEFAULTS['transition_covariance_trend'])
            
            gap_config = kalman_config.get('gap_handling', {})
            gap_variance_multiplier = gap_config.get('gap_variance_multiplier', 2.0)
            trend_variance_multiplier = gap_config.get('trend_variance_multiplier', 20.0)
            
            state['kalman_params']['transition_covariance'] = [
                [base_weight_cov * (1 + gap_factor * gap_variance_multiplier * decay_factor), 0],
                [0, base_trend_cov * (1 + gap_factor * trend_variance_multiplier * decay_factor)]
            ]
            
            adaptation['measurements_since_gap'] += 1
        else:
            state['gap_adaptation']['active'] = False
            kalman_config = config.get('kalman', {})
            state['kalman_params']['transition_covariance'] = [
                [kalman_config.get('transition_covariance_weight', KALMAN_DEFAULTS['transition_covariance_weight']), 0],
                [0, kalman_config.get('transition_covariance_trend', KALMAN_DEFAULTS['transition_covariance_trend'])]
            ]
        
        return state

File: weight-processor-ts/src/core/processing/kalman.ts
/**
 * Kalman filter logic for weight processing.
 * Manages Kalman filter operations including initialization, updates, predictions,
 * and adaptive parameters for post-reset periods.
 */

import { KalmanFilter } from './kalman_filter';
import type { Vector2, Matrix2x2 } from '../math/matrix';
import { KALMAN_DEFAULTS } from '../../constants';
import type {
  ProcessorState,
  KalmanParams,
  ProcessResult,
} from '../../models';

/**
 * Convert value to float, handling nested structures
 */
function ensureFloatFromDecimal(value: any): any {
  if (value === null || value === undefined) {
    return value;
  }
  if (typeof value === 'number') {
    return value;
  }
  if (typeof value === 'string') {
    const parsed = parseFloat(value);
    return isNaN(parsed) ? value : parsed;
  }
  if (Array.isArray(value)) {
    return value.map(v => ensureFloatFromDecimal(v));
  }
  if (typeof value === 'object') {
    const result: any = {};
    for (const [k, v] of Object.entries(value)) {
      result[k] = ensureFloatFromDecimal(v);
    }
    return result;
  }
  return value;
}

export class KalmanFilterManager {
  /**
   * Initialize Kalman filter immediately with first measurement.
   */
  static initializeImmediate(
    weight: number,
    timestamp: Date,
    kalmanConfig: any,
    observationCovariance?: number
  ): Partial<ProcessorState> {
    const initialVariance = kalmanConfig.initial_variance ?? KALMAN_DEFAULTS.initial_variance;

    // Use passed observation_covariance if provided, otherwise use config value
    const obsCov = observationCovariance !== undefined
      ? observationCovariance
      : (kalmanConfig.observation_covariance ?? KALMAN_DEFAULTS.observation_covariance);

    const kalmanParams: KalmanParams = {
      initialStateMean: [weight, 0],
      initialStateCovariance: [[initialVariance, 0], [0, 0.001]],
      transitionCovariance: [
        [
          kalmanConfig.transition_covariance_weight ?? KALMAN_DEFAULTS.transition_covariance_weight,
          0,
        ],
        [
          0,
          kalmanConfig.transition_covariance_trend ?? KALMAN_DEFAULTS.transition_covariance_trend,
        ],
      ],
      observationCovariance: [[obsCov]],
    };

    return {
      kalmanParams,
      lastState: [weight, 0],  // Store as Vector2, not sequence
      lastCovariance: [[initialVariance, 0], [0, 0.001]],  // Store as Matrix2x2, not sequence
      lastTimestamp: timestamp,
      lastRawWeight: weight,
    };
  }

  /**
   * Update Kalman filter state with new measurement.
   */
  static updateState(
    state: ProcessorState,
    weight: number,
    timestamp: Date,
    _source: string,  // Prefix unused params with _
    _processingConfig: any,
    observationCovariance?: number
  ): ProcessorState {
    let timeDeltaDays = 1.0;
    if (state.lastTimestamp) {
      const lastTimestamp = typeof state.lastTimestamp === 'string'
        ? new Date(state.lastTimestamp)
        : state.lastTimestamp;
      const delta = (timestamp.getTime() - lastTimestamp.getTime()) / (86400.0 * 1000);
      timeDeltaDays = Math.max(0.1, Math.min(30.0, delta));
    }

    const kalmanParams = state.kalmanParams!;

    // Use passed observation_covariance if provided, otherwise use stored value
    const obsCov = observationCovariance !== undefined
      ? observationCovariance
      : parseFloat(String(kalmanParams.observationCovariance?.[0]?.[0] ?? 5.0));

    const kalman = new KalmanFilter(
      [[1, timeDeltaDays], [0, 1]],  // transitionMatrix
      [[1, 0], [0, 0]],  // observationMatrix (2x2 with second row zeros)
      ensureFloatFromDecimal(kalmanParams.initialStateMean) as Vector2,
      kalmanParams.initialStateCovariance as Matrix2x2,
      ensureFloatFromDecimal(kalmanParams.transitionCovariance) as Matrix2x2,
      obsCov  // observationCovariance (scalar)
    );

    const observation: number = weight;  // Single observation value

    // Use current state or initialize
    const currentState = state.lastState || kalmanParams.initialStateMean;
    const currentCovariance = state.lastCovariance || kalmanParams.initialStateCovariance;

    const [filteredStateMean, filteredStateCovariance] = kalman.filterUpdate(
      currentState,
      currentCovariance,
      observation
    );

    return {
      ...state,
      lastState: filteredStateMean,
      lastCovariance: filteredStateCovariance,
      lastTimestamp: timestamp,
      lastRawWeight: weight,
    };
  }

  /**
   * Calculate confidence using smooth exponential decay.
   */
  static calculateConfidence(normalizedInnovation: number): number {
    const alpha = 0.5;
    return Math.exp(-alpha * normalizedInnovation ** 2);
  }

  /**
   * Create result dictionary from current state.
   */
  static createResult(
    state: ProcessorState,
    weight: number,
    timestamp: Date,
    source: string,
    accepted: boolean,
    observationCovariance?: number
  ): Partial<ProcessResult> | null {
    if (!state.lastState) {
      return null;
    }

    const currentState = state.lastState;  // Vector2
    const filteredWeight = currentState[0];
    const trend = currentState[1];
    const innovation = weight - filteredWeight;

    const currentCovariance = state.lastCovariance;  // Matrix2x2 or null
    let normalizedInnovation = 0;

    if (currentCovariance) {
      // Use passed observation_covariance if provided, otherwise use stored value
      const obsCovariance = observationCovariance !== undefined
        ? observationCovariance
        : parseFloat(String(state.kalmanParams!.observationCovariance[0]![0]));

      const innovationVariance = currentCovariance[0][0] + obsCovariance;
      normalizedInnovation = innovationVariance > 0
        ? Math.abs(innovation) / Math.sqrt(innovationVariance)
        : 0;
    }

    const confidence = KalmanFilterManager.calculateConfidence(normalizedInnovation);

    // Calculate confidence intervals (±2σ)
    let kalmanUpper: number;
    let kalmanLower: number;
    let kalmanVariance: number | null;

    if (currentCovariance) {
      const confidenceInterval = 2.0 * Math.sqrt(currentCovariance[0][0]);
      kalmanUpper = filteredWeight + confidenceInterval;
      kalmanLower = filteredWeight - confidenceInterval;
      kalmanVariance = currentCovariance[0][0];
    } else {
      kalmanUpper = filteredWeight;
      kalmanLower = filteredWeight;
      kalmanVariance = null;
    }

    // Calculate prediction error
    const predictionError = accepted ? innovation : undefined;

    return {
      timestamp: timestamp.toISOString(),  // Convert Date to string
      rawWeight: weight,
      filteredWeight,
      trend,
      trendWeekly: trend * 7,
      accepted,
      confidence,
      innovation,
      normalizedInnovation,
      source,
      kalmanConfidenceUpper: kalmanUpper,
      kalmanConfidenceLower: kalmanLower,
      kalmanVariance: kalmanVariance ?? undefined,  // Convert null to undefined for optional field
      predictionError,
    };
  }

  /**
   * Extract current weight and trend from state.
   */
  static getCurrentStateValues(state: ProcessorState): [number | null, number | null] {
    if (!state.lastState) {
      return [null, null];
    }

    const currentState = state.lastState;  // Vector2
    return [currentState[0], currentState[1]];
  }

  /**
   * Calculate days between measurements.
   */
  static calculateTimeDeltaDays(
    currentTimestamp: Date,
    lastTimestamp: Date | null | undefined
  ): number {
    if (!lastTimestamp) {
      return 1.0;
    }

    const last = typeof lastTimestamp === 'string'
      ? new Date(lastTimestamp)
      : lastTimestamp;

    const delta = (currentTimestamp.getTime() - last.getTime()) / (86400.0 * 1000);
    return Math.max(0.1, delta);
  }

  /**
   * Get prediction for the next timestamp WITHOUT updating state.
   * This is the true Kalman prediction step, used for quality scoring.
   *
   * Returns [predicted_weight, innovation_covariance] or [null, null]
   */
  static predictNextState(
    state: ProcessorState,
    timestamp: Date,
    _config?: any  // Prefix unused param with _
  ): [number | null, number | null] {
    // Check if we have a valid state
    if (
      !state ||
      !state.lastState ||
      !state.kalmanParams
    ) {
      return [null, null];
    }

    // Get last timestamp
    const lastTimestamp = state.lastTimestamp;
    if (!lastTimestamp) {
      return [null, null];
    }

    // Convert to Date if string
    const lastTimestampDate = typeof lastTimestamp === 'string'
      ? new Date(lastTimestamp)
      : lastTimestamp;

    // Calculate time delta
    const timeDeltaDays = KalmanFilterManager.calculateTimeDeltaDays(
      timestamp,
      lastTimestampDate
    );

    // Get last posterior state and covariance
    const posteriorState = state.lastState;  // Vector2
    const posteriorCovariance = state.lastCovariance!;  // Matrix2x2

    // Build transition matrix F
    const F: number[][] = [[1, timeDeltaDays], [0, 1]];

    // Get process noise Q from kalman_params
    const kalmanParams = state.kalmanParams!;  // Already checked above
    const Q: number[][] = ensureFloatFromDecimal(kalmanParams.transitionCovariance);

    // Predict state: x_pred = F * x_posterior
    const predictedState: number[] = [
      F[0]![0]! * posteriorState[0] + F[0]![1]! * posteriorState[1],
      F[1]![0]! * posteriorState[0] + F[1]![1]! * posteriorState[1],
    ];

    // Predict covariance: P_pred = F * P_posterior * F' + Q
    // First: F * P_posterior
    const FP: number[][] = [
      [
        F[0]![0]! * posteriorCovariance[0]![0]! + F[0]![1]! * posteriorCovariance[1]![0]!,
        F[0]![0]! * posteriorCovariance[0]![1]! + F[0]![1]! * posteriorCovariance[1]![1]!,
      ],
      [
        F[1]![0]! * posteriorCovariance[0]![0]! + F[1]![1]! * posteriorCovariance[1]![0]!,
        F[1]![0]! * posteriorCovariance[0]![1]! + F[1]![1]! * posteriorCovariance[1]![1]!,
      ],
    ];

    // Then: (F * P_posterior) * F'
    const FPFt: number[][] = [
      [
        FP[0]![0]! * F[0]![0]! + FP[0]![1]! * F[0]![1]!,
        FP[0]![0]! * F[1]![0]! + FP[0]![1]! * F[1]![1]!,
      ],
      [
        FP[1]![0]! * F[0]![0]! + FP[1]![1]! * F[0]![1]!,
        FP[1]![0]! * F[1]![0]! + FP[1]![1]! * F[1]![1]!,
      ],
    ];

    // Finally: F * P_posterior * F' + Q
    const predictedCovariance: number[][] = [
      [FPFt[0]![0]! + Q[0]![0]!, FPFt[0]![1]! + Q[0]![1]!],
      [FPFt[1]![0]! + Q[1]![0]!, FPFt[1]![1]! + Q[1]![1]!],
    ];

    // Extract predicted weight (first element of state vector)
    const predictedWeight = predictedState[0] ?? null;

    // Calculate innovation covariance for the measurement
    // S = H * P_pred * H' + R, where H = [1, 0] for weight observation
    // Since H = [1, 0], this simplifies to P_pred[0,0] + R
    const R = parseFloat(String(kalmanParams.observationCovariance[0]![0]!));
    const innovationCovariance = predictedCovariance[0]![0]! + R;

    return [predictedWeight, innovationCovariance];
  }

  /**
   * Get adaptive covariances that start loose after reset and tighten over time.
   * This helps the filter adapt quickly to new weight patterns after a gap.
   */
  static getAdaptiveCovariances(
    measurementsSinceReset: number,
    config: any
  ): { weight: number; trend: number } {
    const baseWeightCov = config.transition_covariance_weight ??
      KALMAN_DEFAULTS.transition_covariance_weight;
    const baseTrendCov = config.transition_covariance_trend ??
      KALMAN_DEFAULTS.transition_covariance_trend;

    // Get adaptive settings from config
    const adaptiveConfig = config.post_reset_adaptation ?? {};

    const warmupMeasurements = adaptiveConfig.warmup_measurements ?? 10;
    const weightBoostFactor = adaptiveConfig.weight_boost_factor ?? 10;
    const trendBoostFactor = adaptiveConfig.trend_boost_factor ?? 100;
    const decayRate = adaptiveConfig.decay_rate ?? 3;

    if (measurementsSinceReset < warmupMeasurements) {
      // Exponentially decay from boost to 1x over warmup period
      const factor = Math.exp(-measurementsSinceReset / decayRate);

      const weightMultiplier = 1 + (weightBoostFactor - 1) * factor;
      const trendMultiplier = 1 + (trendBoostFactor - 1) * factor;

      return {
        weight: baseWeightCov * weightMultiplier,
        trend: baseTrendCov * trendMultiplier,
      };
    } else {
      // After warmup, use normal values
      return { weight: baseWeightCov, trend: baseTrendCov };
    }
  }
}

/**
 * Get adaptive Kalman parameters that gradually transition from
 * loose (adaptive) to tight (normal) configuration after a reset.
 *
 * Uses multipliers from reset parameters to scale base config values.
 */
export function getAdaptiveKalmanParams(
  resetTimestamp: Date | null | undefined,
  currentTimestamp: Date,
  baseConfig: any,
  adaptiveDays: number = 7,
  state?: ProcessorState
): any {
  if (!resetTimestamp) {
    return baseConfig;
  }

  const reset = typeof resetTimestamp === 'string'
    ? new Date(resetTimestamp)
    : resetTimestamp;

  const daysSinceReset = (currentTimestamp.getTime() - reset.getTime()) / (86400.0 * 1000);

  // Initialize adaptation_days (may be overridden by reset_parameters)
  let adaptationDaysValue = adaptiveDays;
  let resetParams: any = {}; // Default empty reset params

  // Check if we have custom reset parameters in state
  if (state && state.resetParameters) {
    resetParams = state.resetParameters;
    adaptationDaysValue = resetParams.adaptation_days ?? adaptiveDays;

    // Get multipliers from reset parameters
    const initialVarMult = resetParams.initial_variance_multiplier ?? 5;
    const weightNoiseMult = resetParams.weight_noise_multiplier ?? 20;
    const trendNoiseMult = resetParams.trend_noise_multiplier ?? 200;
    const obsNoiseMult = resetParams.observation_noise_multiplier ?? 0.5;

    // Apply multipliers to base config
    const baseInitialVar = baseConfig.initial_variance ?? 0.361;
    const baseWeightCov = baseConfig.transition_covariance_weight ?? 0.016;
    const baseTrendCov = baseConfig.transition_covariance_trend ?? 0.0001;
    const baseObsCov = baseConfig.observation_covariance ?? 3.4;

    const adaptiveParams = {
      initial_variance: baseInitialVar * initialVarMult,
      transition_covariance_weight: baseWeightCov * weightNoiseMult,
      transition_covariance_trend: baseTrendCov * trendNoiseMult,
      observation_covariance: baseObsCov * obsNoiseMult,
    };

    // Check if we're still in adaptation period
    if (daysSinceReset >= adaptationDaysValue) {
      return baseConfig;
    }

    // Calculate decay factor based on days since reset
    const decayRate = resetParams.adaptation_decay_rate ?? 2.5;
    const measurementsSince = state?.measurementsSinceReset ?? 0;

    // Use measurement-based decay if available, otherwise time-based
    let decayFactor: number;
    if (measurementsSince > 0) {
      decayFactor = 1.0 - Math.exp(-measurementsSince / decayRate);
    } else {
      decayFactor = Math.min(1.0, daysSinceReset / adaptationDaysValue);
    }

    // Interpolate between adaptive and base parameters
    const result: any = {};
    for (const key of Object.keys(baseConfig)) {
      if (key in adaptiveParams) {
        const adaptiveValue = (adaptiveParams as any)[key];
        const baseValue = baseConfig[key];
        result[key] = adaptiveValue + decayFactor * (baseValue - adaptiveValue);
      } else {
        result[key] = baseConfig[key];
      }
    }

    return result;
  } else {
    // Use default adaptive parameters (shouldn't happen with proper reset)
    const adaptiveParams = {
      initial_variance: 5.0,
      transition_covariance_weight: 0.5,
      transition_covariance_trend: 0.01,
      observation_covariance: 2.0,
    };

    // Check if we're still in adaptation period
    if (daysSinceReset >= adaptationDaysValue) {
      return baseConfig;
    }

    // Simple time-based decay
    const decayFactor = Math.min(1.0, daysSinceReset / adaptationDaysValue);

    // Interpolate between adaptive and base parameters
    const result: any = {};
    for (const key of Object.keys(baseConfig)) {
      if (key in adaptiveParams) {
        const adaptiveValue = (adaptiveParams as any)[key];
        const baseValue = baseConfig[key];
        result[key] = adaptiveValue + decayFactor * (baseValue - adaptiveValue);
      } else {
        result[key] = baseConfig[key];
      }
    }

    return result;
  }
}
